src/hex_tarball.erl

```diff
@@ -0,0 +1,455 @@
+-module(hex_tarball).
+-export([create/2, unpack/2, format_checksum/1, format_error/1, gzip/1, normalize_requirements/1]).
+-define(VERSION, <<"3">>).
+-define(TARBALL_MAX_SIZE, 8 * 1024 * 1024).
+-define(BUILD_TOOL_FILES, [
+                            {<<"mix.exs">>, "mix"},
+                            {<<"rebar.config">>, <<"rebar">>},
+                            {<<"rebar">>, <<"rebar">>},
+                            {<<"Makefile">>, <<"make">>},
+                            {<<"Makefile.win">>, <<"make">>}
+                           ]).
+-include_lib("kernel/include/file.hrl").
+
+-type checksum() :: binary().
+-type contents() :: #{filename() => binary()}.
+-type filename() :: string().
+-type files() :: [filename()] | contents().
+-type metadata() :: map().
```

Comment:
currently metadata has binary keys, eventually I'd like to have atom keys but not in this PR as it requires more changes in hexpm and has security implications.